### ISSUE TITLE
confusing batch_normalization

### ISSUE THREAD

**Issue by jiefengpeng** [ppwwyyxx/tensorpack#457]
Hi, when I use nn.moments to calculate the mean & variance and use nn.batch_normalization to output xn instead of nn.fused_batch_norm, the training speed decresses from 1.8iter/s to 0.3iter/s. Could you figure out what happen to nn.moments with nn.batch_normalization?
I try this in other code, and it seems normal.

**Comment by ppwwyyxx** [comment]
fused_batch_norm is faster than batch_normalization.

This is not a tensorpack question. closing.

**Comment by jiefengpeng** [comment]
But many other platforms implementing ResNets use tf.moments + batch_normalization instead of fused_batch_norm, such as the code training from scratch in Tensorflow official webpage, their speed is not so slow. Any communication delay problem in your code when using tf.moments + batch_normalization?

**Comment by ppwwyyxx** [comment]
No, they are slow. https://www.tensorflow.org/performance/performance_models#use_fused_batch-normalization

**Comment by jiefengpeng** [comment]
Ok, I got it that fused_batch_norm is faster than batch_normalization, but my point is that in other resnet implemented framework, which use moment + batch_normalization, their speed is 2 more times faster than your code with moment + batch_normalization.

**Comment by jiefengpeng** [comment]
By the way, we train on 8 TITAN X.

**Comment by ppwwyyxx** [comment]
Assuming you're talking about single GPU training, and some thing runs significantly faster with other code. The only way that explains this is that they are not the same model with the same configuration. You'll need to double check batch size, data layout, etc, and make sure non-model factors (e.g. data loading) is not affecting your observation.

**Comment by ppwwyyxx** [comment]
"tensorpack slower than others" is a potential performance issue then. To get more serious you'd better also include what you did (command and settings), what you observed (logs and speed numbers), if possible.

**Comment by ppwwyyxx** [comment]
One possible explanation, for example, is that non-fused batch norm is faster with NHWC (according to https://github.com/tensorflow/tensorflow/issues/12419) than NCHW. Our resnet uses fused batch norm and therefore NCHW. If the other code uses NHWC, then you're not comparing apple-to-apple.

**Comment by jiefengpeng** [comment]
Ah, actually it is multi-GPU training with 8 TITAN X. And we've checked that the batch size and data layout are as same as yours. The only difference is they use moving_mean and moving_var in training for which they need to use moment&batch_normalization. I'm really confused that fused_batch_norm is significantly 6 more times faster than moment&batch_normalization. What's more, we observed a poor GPU utilization with moment&batch_normalization which is around 100% when using fused_batch_norm.

**Comment by ppwwyyxx** [comment]
> The only difference is they use moving_mean and moving_var in training for which they need to use moment&batch_normalization
I don't understand from this sentence what the difference is. AFAIK nn.moments doesn't give you moving_mean and moving_var.

As I said, if you want this to be an issue about "tensorpack slower than others", I would need much more details.

**Comment by jiefengpeng** [comment]
With nn.moments they calculate bm and bv, and update bm and bv with moving_mean and moving_var, after which they are fed to non-fused batch norm to output xn in training.
Ok, we try to collect details later and thanks a lot to your patience.

**Comment by ppwwyyxx** [comment]
> With nn.moments they calculate bm and bv, and update bm and bv with moving_mean and moving_var, after which they are fed to non-fused batch norm to output xn in training.

Thanks. This looks like standard (if not the only) way to use non-fused BN. How do you do it differently in tensorpack?

**Comment by jiefengpeng** [comment]
Actually we try to do something like renormalization in tensorpack.

**Comment by jiefengpeng** [comment]
As you said, the reason is the data format with NCHW. https://github.com/tensorflow/tensorflow/issues/9141
Is there any problem occurred if we transform the data format at data layer to NHWC in tensorpack?

**Comment by ppwwyyxx** [comment]
You said that the "other implementation" use the same data layout, right?
The issue is "tensorpack vs the other implementation", not "fused vs non-fused", because non-fused is slow as expected.

The resnet example in tensorpack support both data layout.

**Comment by jiefengpeng** [comment]
Sorry! I mistake the data layout with other implementation.
So there is no issue of "tensorpack vs the other implementation", and thanks again for your kindness!
